File: clienttier/controller/actions/BuildConstruction.js
/**
 * FILE: BuildConstruction.js
 * PERCORSO /Codice/sgad/clienttier/controller/actions/BuildConstruction.js
 * DATA CREAZIONE: 22 Febbraio 2014
 * AUTORE: ProTech
 * EMAIL: protech.unipd@gmail.com
 *
 * Questo file è proprietà del gruppo ProTech, viene rilasciato sotto licenza Apache v2.
 *
 * DIARIO DELLE MODIFICHE:
 * 2014-02-22 - Creazione della classe - Battistella Stefano
 */

//Eredita da Action
BuildConstruction.prototype = new Action();

//Costruttore della classe
BuildConstruction.prototype.constructor = BuildConstruction;

/**
 * Classe per la gestione della costruzione di un nuovo edificio nel villaggio dell'utente.
 * @implements Action
 * @constructor
 * @param {TileComponent} tileComponent La casella della mappa dove costruire l'edificio.
 * @param {string} key Tipo di edificio da rappresentare.
 */
function BuildConstruction(tileComponent, key) {
    /**
     * tileComponent La casella della mappa dove costruire l'edificio.
     * @type {TileComponent}
     * @private
     */
    this.tileComponent = tileComponent;
    /**
     * Tipo di edificio da rappresentare.
     * @type {string}
     * @private
     */
    this.key = key;
    /**
     * Dati relativi all'edificio da costruire.
     * @type {Object}
     * @private
     */
    this.datas = null;
}

/**
 * Esegue la lista di azioni necessaria per costruire il nuovo edificio.
 * @override
 * @return {void}
 */
BuildConstruction.prototype.performAction = function() {
	if(this.datas === null) { //controllo se l'operazione è fallita
		var operationFailureMenu = new ShowOperationFailureMenu(); //visualizzo il fallimento dell'operazione
		operationFailureMenu.performAction();
		return;
	}
	var context = Context.getInstance();
	if(!this.datas) {
		var menu = (new NotifyMenuFactory("Attenzione! Non hai sufficienti lavoratori!.", null)).buildMenu();
		var widgets = menu.getWidgets(); //ottengo i widgets memorizzati nel frame
		for (var j = 0; j < widgets.length; j++)
			context.addGraphicObject(widgets[j]); //aggiungo i widget per poterli visualizzare
		context.addGraphicObject(menu); //aggiungo il frame
		return;
	}

	//l'operazione è andata a buon fine
	context.removeGraphicObject(this.tileComponent); //rimuovo la casella dove deve essere costruito l'edificio

	var userData = UserDataManager.getInstance().getUserData();
	var dataFactory = DataFactory.getInstance();
	var building = dataFactory.getBuilding(this.key);//ottengo il tipo di costruzione a cui associare l'edificio

	var cost = building.getCost();//sottraggo i costi della costruzione
	var quantityResources = cost.getQuantityResource();
	for(var i = 0; i < quantityResources.length; i++)
	{
		var key = quantityResources[i].getResource().getKey();
		var ownedResource = userData.getOwnedResource(key);
		ownedResource.setQuantity(ownedResource.getQuantity() - quantityResources[i].getQuantity());
	}

	var position = this.tileComponent.getPosition(); //ottengo la posizione in cui piazzare l'edificio
	var time = building.getCost().getRelativeTime(); //ottengo il tempo necessario a costruire l'edificio
	var buildingPossession = new BuildingPossession(building, false, position, time, 0, null); //creo l'edificio del modello
    buildingPossession.setUserData(userData);
	var buildingComponent = new BuildingComponent(buildingPossession, this.key); //creo l'edificio da visualizzare
	buildingComponent.onClickAction = new RequestHarvestResources(buildingPossession);
	buildingComponent.onRightClickAction = new ShowBuildingContextualMenu(buildingComponent);
	buildingPossession.addObserver(buildingComponent);
	var setAllVoidFilter = new SetAllVoidFilter(); //imposto che il filtro una volta costruito sia quello nullo
	userData.addBuildingPossession(buildingPossession);
	setAllVoidFilter.performAction();
	new VoidFilter(buildingComponent);
	context.addGraphicObject(buildingComponent); //aggiungo l'oggetto grafico
};

/**
 * Imposta i dati relativi all'edificio da costruire.
 * @param {Object} datas I dati da impostare.
 * @return {void}
 */
BuildConstruction.prototype.setActionDatas = function(datas) {
	this.datas = datas;
};

File: Codice/servertier/src/main/scala/sgad/servertier/businesslogic/operations/BuildConstruction.scala
/**
 * FILE: BuildConstruction.scala
 * PERCORSO /Codice/sgad/servertier/src/main/scala/sgad/servertier/businesslogic/operations
 * DATA CREAZIONE: 25 Febbraio 2014
 * AUTORE: ProTech
 * EMAIL: protech.unipd@gmail.com
 *
 * Questo file è proprietà del gruppo ProTech, viene rilasciato sotto licenza Apache v2.
 *
 * DIARIO DELLE MODIFICHE:
 * 2014-02-25 - Creazione della classe - Segantin Fabio
 */

package sgad.servertier.businesslogic.operations

import sgad.servertier.dataaccess.data.userdata.{Position, BuildingPossession, UserData}
import sgad.servertier.dataaccess.data.shareddata.{BuildingWithLevel, DataFactory}
import java.util.NoSuchElementException

/**
 * Classe che rappresenta l'operazione di costruzione di un edificio.
 */
class BuildConstruction extends Operation {
	/**
	 * Esecuzione della costruzione.
	 * @param userData Dati dell'utente su cui verrà effettuata l'operazione.
	 * @param data Dati accompagnatori alla richiesta dell'operazione.
	 * @param loginAuthorization Autorizzazione a operare richieste di login. Di defalut è false.
	 * @param registrationAuthorization Autorizzazione a operare richieste di registrazione. Di defalut è false.
	 * @param userAuthorization Autorizzazione a operare richieste di user. Di defalut è false.
	 * @param internalAuthorization Autorizzazione a operare richieste interne. Di defalut è false.
	 * @return Stringa contenente la risposta.
	 */
	def execute(userData: UserData, data: String, loginAuthorization: Boolean, registrationAuthorization: Boolean,
	            userAuthorization: Boolean, internalAuthorization: Boolean): String = {

		OperationFactory.getOperation("UpdateUserData").execute(userData, "", internalAuthorization = true)
		var answer = "data:false, unauthorized:true"
		if (userAuthorization) {
			//viene controllata l'autorizzazione
			val dataMap = decodeData(data)
			try {
				if (userData.getAuthenticationData.getAuthenticationString == dataMap("authentication")) {
					//controllata la stringa di autenticazione
					val newPosition = new Position(dataMap("x").toInt, dataMap("y").toInt)
					val existanceCheck = userData.getOwnedBuildingsMap.values.forall((building: BuildingPossession) => {
						building.getPosition != newPosition
					}) //controllo non sovrapposizione di edifici
					if (existanceCheck) {
						val objectBuilding = DataFactory.getBuilding(dataMap("key"))
						if (objectBuilding.getIsConstructible) {
							//viene controllato che l'edificio sia costruibile per statistica.
							val preconditionCheck = objectBuilding.getPrecondition.forall((building) => {
								userData.buildingIsOwned(building.asInstanceOf[BuildingWithLevel].getKey)
							}) //controllo le precondizioni
							val resourceCheck = objectBuilding.getCost.getQuantityResource.forall((resource) => {
									userData.getOwnedResource(resource.getResource.getKey).getQuantity >= resource.getQuantity
								}) //controllo se possiede le risorse
							val workersavaible = userData.getOwnedUnitMap.values.foldLeft(0)((current, unit) => {
									//sommo tutti gli elementi conenuti nella mappa delle unità possedute solo se sono lavoratori
									var count = current
									if (unit.getUnit.getIsBuilder)
										count += unit.getQuantity
									count
								}) - userData.getOwnedBuildingsMap.values.foldLeft(0)((current: Int, building) => {
									//per ogni edificio posseduto sommo 1 se non è finito
									var count = current
									if (!building.getIsFinished)
										count += 1
									count
								})
							//ottengo il numero di lavoratori che non stanno eseguendo un lavoro
							if (preconditionCheck && resourceCheck && workersavaible > 0) {
								//gli edifici richiesti, le risorse richieste e i lavoratori disponibili permettono la costruzione dell'edificio in questione
								build(objectBuilding, userData, newPosition)
								answer = "data:true, messages:" + parsePiggy(userData.getPiggy)
							} else
								//non possiede risorse, precondizioni o lavoratori a sufficienza
								answer = "data:false, condition:false, messages:" + parsePiggy(userData.getPiggy)
						} else
							//l'edificio non è costruibile
							answer = "data:false, costructible:false, messages:" + parsePiggy(userData.getPiggy)
					} else
						//l'edificio non può essere costruito sopra ad un altro edificio.
						answer = "data:false, position:false, messages:" + parsePiggy(userData.getPiggy)
				} else
					//l'autenticazione della richiesta non è andata a buon fine.
					answer = "data:false, authentication:false"
			} catch {
				case _: IllegalArgumentException => answer = "data:false, parameters:false" //C'è stato un errore nella costruzione di un elemento
				case _: ClassCastException => answer = "data:false, parameters:false" //C'è stato un errore durante la parserizzazione di un elemento
				case _: NoSuchElementException => answer = "data:false, parameters:false" //l'elemento non è posseduto e non esiste nel sistema.
			}
		}
		"{" + answer + "}" //ritorno la risposta
	}

	/**
	 * Metodo utilizzato dalla classe per effettuare la costruzione, il metodo richiede che i test di consistenza con i dati siano stati eseguiti.
	 * @param objectBuilding È la BuildingWithLevel da costruire.
	 * @param userData I dati utente di chi richieda la costruzione.
	 * @param newPosition È la posizione in cui si costruisce la costruzione.
	 * @return Unit
	 */
	private def build(objectBuilding: BuildingWithLevel, userData: UserData, newPosition: Position) = {
		//i controlli di consistenza sono già stati eseguiti
		//la successiva rimozione di risorse prevede di non settare le risorse a meno di 0.
		objectBuilding.getCost.getQuantityResource.foreach((resource) => {
			//riduco le risorse possedute
			val ownedResource = userData.getOwnedResource(resource.getResource.getKey)
			ownedResource.setQuantity(ownedResource.getQuantity - resource.getQuantity)
		})
		userData.addBuildingPossession(//aggiungo
			//avendo già controllato la posizione e i requisiti sono sicuro di non creare situazioni di inconsistenza
			new BuildingPossession(
				objectBuilding,
				newPosition,
				isFinished = false,
				time = System.currentTimeMillis() / 1000L,
				null
			)
		)
	}
}
